### PR TITLE
Ripley stats for categorical variables

### PR DESCRIPTION
**Pull request by giovp**
### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (change that would cause existing functionality to not work as before)
- [x] New feature (non-breaking change which adds functionality)

### Description

Implement various ripley's alphabet statistics for point patterns analysis.
Will also move stuff around a bit on the `_patterns` module

### How has this been tested?

...

### Closes

...

### CODE REVIEW

**Comment by giovp** [comment]
Ok, ready to review, best to be tested with the lymphnode dataset that has some spatial patterns for clusters

<details>
<summary>Details</summary>

```python
import scanpy as sc
import squidpy as sq

adata = sc.datasets.visium_sge("V1_Human_Lymph_Node")
sc.pp.normalize_total(adata)
sc.pp.log1p(adata)
sc.pp.highly_variable_genes(adata)
sc.pp.pca(adata)
sc.pp.neighbors(adata)
sc.tl.leiden(adata)
sc.tl.umap(adata)
sc.pl.umap(adata, color="leiden")
sc.pl.spatial(adata, color="leiden")

mode = "L"
sq.gr.ripley(adata, cluster_key="leiden", mode=mode)
sq.pl.ripley(adata, cluster_key="leiden", mode=mode)
```

image

image


</details>

What's now available is ripley F,G and L. L is like K (which we used to have before) but `sqrt(K/pi)` normalized (preferrable in practice). F and G give the sme results for this example, but if the convex hull of the tissue has "holes" it would proabbly be different (essentially G would account for that, F not cause it alwasy compute the score wrt a spatial point process.

It is not optimized but I abstracted the core functions enough to make it flexible. For sure there is room for improvement but can be done later, first let's see if people use.

**Comment by giovp** [comment]
doics seem to be failing for napari (weird)

**Comment by michalk8** [comment]
> doics seem to be failing for napari (weird)

Strange, neither `nbformat`, nor `nbsphinx` had a new release recently afaik, will look into this and review this.

File: squidpy/gr/_ripley.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+
+__all__ = ["ripley"]
+
+
+@d.dedent
+@inject_docs(key=Key.obsm.spatial, rp=RipleyStat)
+def ripley(
+    adata: AnnData,
+    cluster_key: str,
+    mode: Literal["F", "G", "L"],
```
Comment (michalk8) [inline squidpy/gr/_ripley.py:29]: I'd provide a default here for bwd. compat/ease o use.

File: squidpy/gr/_ripley.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+
+        K(t) = \frac{{1}}{{\lambda}} \sum_{{i \ne j}} \frac{{I(d_{{i,j}}<t)}}{{n}}
+
+    and then we apply a variance-stabilizying transformation:
+
+    .. math::
+
+        L(t) = (\frac{{K(t)}}{{\pi}})^{{1/2}}
+
+    For reference, check out
```
Comment (michalk8) [inline squidpy/gr/_ripley.py:69]: Please move this to 
```
References
----------------
```
section below `Returns`.

File: squidpy/gr/_ripley.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    cluster_idx = le.transform(clusters)
+    obs_arr = np.empty((le.classes_.shape[0], n_steps))
+
+    start = logg.info(
+        f"Calculating Ripley's {mode} statistic for `{le.classes_.shape[0]}` clusters and `{n_simulations}` simulations"
+    )
+
+    for i in np.arange(np.max(cluster_idx) + 1):
+        coord_c = coordinates[cluster_idx == i, :]
+        if mode == "F":
```
Comment (michalk8) [inline squidpy/gr/_ripley.py:128]: ~~You're comparing an enum to string, can be dangerous if the impl. of `ModeEnum` changes in the future.
Please use `mode == RipleyStat.F`, etc.~~
My bad, you're just assigning `mode = mode.s` right after. However, I wouldn't do that and would just compare the enums, because if you change the value in the enum, PyCharm/your editor of preference won't most likely pick it up.

File: squidpy/gr/_ripley.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+            tree_c = NearestNeighbors(metric=metric, n_neighbors=n_neigh).fit(coord_c)
+            distances, _ = tree_c.kneighbors(random, n_neighbors=n_neigh)
+            bins, obs_stats = _f_g_function(distances.squeeze(), support)
+        elif mode == "G":
+            tree_c = NearestNeighbors(metric=metric, n_neighbors=n_neigh).fit(coord_c)
+            distances, _ = tree_c.kneighbors(coordinates[cluster_idx != i, :], n_neighbors=n_neigh)
+            bins, obs_stats = _f_g_function(distances.squeeze(), support)
+        elif mode == "L":
+            distances = pdist(coord_c, metric=metric)
+            bins, obs_stats = _l_function(distances, support, N, area)
```
Comment (michalk8) [inline squidpy/gr/_ripley.py:139]: I'd add:
```python
else:
   raise NotImplementedError(f"Mode `{mode.s!r}` is not yet implemented.")
```
This prevents cryptic messages if new mode is added/removed and forgotten to be implemented/removed.

File: squidpy/gr/_ripley.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+            tree_i = NearestNeighbors(metric=metric, n_neighbors=n_neigh).fit(random_i)
+            distances_i, _ = tree_i.kneighbors(random, n_neighbors=1)
+            _, stats_i = _f_g_function(distances_i.squeeze(), support)
+        elif mode == "G":
+            tree_i = NearestNeighbors(metric=metric, n_neighbors=n_neigh).fit(random_i)
+            distances_i, _ = tree_i.kneighbors(coordinates, n_neighbors=1)
+            _, stats_i = _f_g_function(distances_i.squeeze(), support)
+        elif mode == "L":
+            distances_i = pdist(random_i, metric=metric)
+            _, stats_i = _l_function(distances_i, support, N, area)
```
Comment (michalk8) [inline squidpy/gr/_ripley.py:157]: Same `else` statement (NotImplemetedError).

File: squidpy/pl/_graph.py [hunk 6]
```diff
@@ -256,9 +256,11 @@ def nhood_enrichment(
 
 
 @d.dedent
-def ripley_k(
+def ripley(
     adata: AnnData,
     cluster_key: str,
+    mode: Literal["F", "G", "L"],
```
Comment (michalk8) [inline squidpy/pl/_graph.py:262]: Again, I'd add default.

File: squidpy/gr/_ripley.py [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+    n_observations: int = 1000,
+    max_dist: np.float_ = None,
+    n_steps: int = 50,
+    seed: Optional[int] = None,
+    copy: bool = False,
+) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+    r"""
+    Calculate various Ripley's statistics for point processes.
+
+    According to the `'mode'` argument, it calculates either the following Ripley's statistics:
```
Comment (michalk8) [inline squidpy/gr/_ripley.py:43]: I'd rephrase as `... it calculates one of the ...`.

File: squidpy/gr/_ripley.py [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+
+    `{rp.F.s!r}`, `{rp.G.s!r}` are defined as:
+
+    .. math::
+
+        F(t),G(t)=P( d_{{i,j}} \le t )
+
+    Where :math:`d_{{i,j}}` represents:
+
+        - distances to a random Spatial Poisson Point Process for `'F'`.
```
Comment (michalk8) [inline squidpy/gr/_ripley.py:54]: You could also use `{rp.F.s!r}` here.

File: squidpy/_docs.py [hunk 9]
```diff
@@ -141,6 +141,19 @@ def decorator2(obj: Any) -> Any:
...
+    - `'{mode}_stat'` - :class:`pandas.DataFrame` containing the statistics of choice for the real observations.
+    - `'sims_stat'` - :class:`pandas.DataFrame` containing the statistics of choice for the simulations.
+    - `'bins'` - :class:`numpy.ndarray` containing the support.
+    - `'pvalues'` - :class:`numpy.ndarray` containing the pvalues for the statistics of interest.
+
+Otherwise, modifies the ``adata`` object with the following key:
+
+    - :attr:`anndata.AnnData.uns` ``['{key_added}']`` - the above mentioned :class:`dict`.
+
+Statistics and pvalues are computed for each cluster `'{cluster_key}'` separately."""
```
Comment (michalk8) [inline squidpy/_docs.py:156]: I'd either rop the `'{cluster_key}'` or use:
... for each cluster in :attr:`anndata.AnnData.obs` ``['{cluster_key}']`` ....

File: squidpy/gr/_ripley.py [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+) -> pd.DataFrame:
+    df = pd.DataFrame(results, columns=columns, index=index)
+    df.index.set_names(["bins"], inplace=True)
+    df = df.melt(var_name=var_name, value_name="stats", ignore_index=False)
+    df[var_name] = df[var_name].astype("category", copy=True)
+    df.reset_index(inplace=True)
+    return df
+
+
+def _f_g_function(distances: np.ndarray, support: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
```
Comment (michalk8) [inline squidpy/gr/_ripley.py:192]: Seems like this could be numba JITted. Maybe leads to perf. gains.

File: squidpy/gr/_ripley.py [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+    counts, bins = np.histogram(distances, bins=support)
+    fracs = np.cumsum(counts) / counts.sum()
+    return bins, np.concatenate((np.zeros((1,), dtype=np.float_), fracs))
+
+
+def _l_function(
```
Comment (michalk8) [inline squidpy/gr/_ripley.py:198]: Same as above, think could be JITted with numba.

**Comment by giovp** [comment]
ah damn missed literal sorry for that. Now looks good! probably temporary bug

File: squidpy/gr/_ripley.py [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+    spatial_key: str = Key.obsm.spatial,
+    metric: str = "euclidean",
+    n_neigh: int = 2,
+    n_simulations: int = 100,
+    n_observations: int = 1000,
+    max_dist: np.float_ = None,
+    n_steps: int = 50,
+    seed: Optional[int] = None,
+    copy: bool = False,
+) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
```
Comment (michalk8) [inline squidpy/gr/_ripley.py:39]: Return type is wrong, should be dict.

File: squidpy/gr/_ripley.py [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+    obs_arr = np.empty((le.classes_.shape[0], n_steps))
+
+    start = logg.info(
+        f"Calculating Ripley's {mode} statistic for `{le.classes_.shape[0]}` clusters and `{n_simulations}` simulations"
+    )
+
+    for i in np.arange(np.max(cluster_idx) + 1):
+        coord_c = coordinates[cluster_idx == i, :]
+        if mode == "F":
+            random = _ppp(hull, 1, n_observations, seed=seed)
```
Comment (michalk8) [inline squidpy/gr/_ripley.py:129]: I'd remove the magic constant (1) as `n_simulations=1` (better readability).

File: squidpy/gr/_ripley.py [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+            bins, obs_stats = _f_g_function(distances.squeeze(), support)
+        elif mode == "L":
+            distances = pdist(coord_c, metric=metric)
+            bins, obs_stats = _l_function(distances, support, N, area)
+        obs_arr[i] = obs_stats
+
+    sims = np.empty((n_simulations, len(bins)))
+    pvalues = np.ones((le.classes_.shape[0], len(bins)))
+
+    for i in range(n_simulations):
```
Comment (michalk8) [inline squidpy/gr/_ripley.py:145]: Prime target for `_parallelize`. I know you've said optimize later and if you want, I can just add it once the other comments are addressed (shouldn't take long).

File: squidpy/gr/_ripley.py [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+    n_simulations
+        Number of simulated point processes.
+    n_observations
+        Number of observations to sample from each simulation
+    seed
+        Random seed.
+
+    Returns
+    -------
+        An Array with shape (n_simulation, n_observations, 2).
```
Comment (michalk8) [inline squidpy/gr/_ripley.py:225]: Unnecessary indentation + please wrap the shape in double backticks as \`\`(n_simulation, n_observations, 2)\`\`.

File: squidpy/gr/_ripley.py [hunk 16]
```diff
@@ -0,0 +1,245 @@
...
+    Simulate Poisson Point Process on a polygon.
+
+    Parameters
+    ----------
+    hull
+        Convex hull of the area of interest.
+    n_simulations
+        Number of simulated point processes.
+    n_observations
+        Number of observations to sample from each simulation
```
Comment (michalk8) [inline squidpy/gr/_ripley.py:219]: Missing `.` at the end.

File: squidpy/gr/_ripley.py [hunk 17]
```diff
@@ -0,0 +1,245 @@
...
+        Which Ripley's statistic to compute.
+    %(spatial_key)s
+    metric
+        Which metric to use for computing distances.
+    n_neigh
+        Number of neighborhoods to consider for the knn graph.
+    n_simulations
+        How many simulations to run for computing pvalues.
+    n_observations
+        How many observations to generate for the spatial poisson point process.
```
Comment (michalk8) [inline squidpy/gr/_ripley.py:87]: Would capitalize as `spatial Poisson point process`  (personal preference).

File: squidpy/gr/_ripley.py [hunk 18]
```diff
@@ -0,0 +1,245 @@
...
+    ----------
+    %(adata)s
+    %(cluster_key)s
+    mode
+        Which Ripley's statistic to compute.
+    %(spatial_key)s
+    metric
+        Which metric to use for computing distances.
+    n_neigh
+        Number of neighborhoods to consider for the knn graph.
```
Comment (michalk8) [inline squidpy/gr/_ripley.py:83]: Would use `KNN` instead of `knn` (personal preference).

File: squidpy/gr/_ripley.py [hunk 19]
```diff
@@ -0,0 +1,245 @@
...
+
+    Parameters
+    ----------
+    %(adata)s
+    %(cluster_key)s
+    mode
+        Which Ripley's statistic to compute.
+    %(spatial_key)s
+    metric
+        Which metric to use for computing distances.
```
Comment (michalk8) [inline squidpy/gr/_ripley.py:81]: I'd add reference as it's done in sklearn for the available metrics: https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestNeighbors.html#sklearn-neighbors-nearestneighbors

File: squidpy/gr/_ripley.py [hunk 20]
```diff
@@ -0,0 +1,245 @@
...
+    res = {f"{mode}_stat": obs_df, "sims_stat": sims_df, "bins": bins, "pvalues": pvalues}
+
+    if TYPE_CHECKING:
+        assert isinstance(res, dict)
+
+    if copy:
+        logg.info("Finish", time=start)
+        return res
+
+    _save_data(adata, attr="uns", key=Key.uns.ripley(cluster_key, mode), data=res)
```
Comment (michalk8) [inline squidpy/gr/_ripley.py:178]: Missing `time=start` in the call.

**Review by michalk8 — CHANGES_REQUESTED** [review]

File: squidpy/gr/_ripley.py (hunk 19 above)
Comment (giovp) [reply squidpy/gr/_ripley.py:81]: added direct link, potentially we can add scikit learn as intersphinx mapping? let me know what you think

File: squidpy/gr/_ripley.py (hunk 10 above)
Comment (giovp) [reply squidpy/gr/_ripley.py:199]: experimented a bit an decided to go against it, as there are no clear performance gains
image

File: squidpy/gr/_ripley.py (hunk 14 above)
Comment (giovp) [reply squidpy/gr/_ripley.py:150]: yes but I feel that there will be a re-write of this in pure numba at some point (was contacted by someone who was interested in doing this). Since for the practical data at hand is not super needed right now (as it takes few seconds anyway, I think we can postpone it. And again like you said, would be straightforwad to add. Thanks for suggesting!

File: squidpy/gr/_ripley.py (hunk 11 above)
Comment (giovp) [reply squidpy/gr/_ripley.py:205]: same as above

File: squidpy/gr/_ripley.py (hunk 19 above)
Comment (michalk8) [reply squidpy/gr/_ripley.py:77]: Yes, it should be added as `sklearn=("https://scikit-learn.org/stable/", None)`.

File: squidpy/gr/_ripley.py (hunk 14 above)
Comment (michalk8) [reply squidpy/gr/_ripley.py:150]: Ok, agreed. After this is merged, could you please create an issue for this (maybe mark it as low-prio) so that it isn't forgotten?

**Comment by giovp** [comment]
must say was a very gentle review 😅 😅 thanks ! Addressed pretty much everything like you suggested, ready to review again.

**Review by michalk8 — APPROVED** [review]

**Comment by michalk8** [comment]
Thanks a lot @giovp , really nice PR, you can go ahead and merged (I've only fixed some typo + link to neighbors).
